data/transactions/logic/pairing.go

```diff
@@ -18,98 +18,766 @@ package logic
+// for product of scalar and point
+func opEcScalarMul(cx *EvalContext) error {
+	group := EcGroup(cx.program[cx.pc+1])
+	fs, ok := ecGroupSpecByField(group)
+	if !ok { // no version check yet, both appeared at once
+		return fmt.Errorf("invalid ec_scalar_mul group %s", group)
+	}
 
-func opBn256Add(cx *EvalContext) error {
 	last := len(cx.stack) - 1
 	prev := last - 1
 	aBytes := cx.stack[prev].Bytes
-	bBytes := cx.stack[last].Bytes
-	if len(aBytes) != 64 || len(bBytes) != 64 {
-		return errors.New("expect G1 in 64 bytes")
-	}
-	a := bytesToBN254G1(aBytes)
-	b := bytesToBN254G1(bBytes)
-	res := new(bn254.G1Affine).Add(&a, &b)
-	resBytes := bN254G1ToBytes(res)
+	kBytes := cx.stack[last].Bytes
+	if len(kBytes) > 32 {
+		return fmt.Errorf("ec_scalar_mul scalar len is %d, exceeds 32", len(kBytes))
+	}
+	k := new(big.Int).SetBytes(kBytes)
+
+	var res []byte
+	var err error
+	switch fs.field {
+	case BN254g1:
+		res, err = bn254G1ScalarMul(aBytes, k)
+	case BN254g2:
+		res, err = bn254G2ScalarMul(aBytes, k)
+	case BLS12_381g1:
+		res, err = bls12381G1ScalarMul(aBytes, k)
+	case BLS12_381g2:
+		res, err = bls12381G2ScalarMul(aBytes, k)
+	default:
+		err = fmt.Errorf("invalid ec_scalar_mul group %s", group)
+	}
+
 	cx.stack = cx.stack[:last]
-	cx.stack[prev].Bytes = resBytes
-	return nil
+	cx.stack[prev].Bytes = res
+	return err
 }
 
-func opBn256ScalarMul(cx *EvalContext) error {
+// Input: Two byte slices, top is concatenated uncompressed bytes for k g2 points, and second to top is same for g1
+// Output: Single uint at top representing bool for whether pairing of inputs was identity
+func opEcPairingCheck(cx *EvalContext) error {
+	/*
+	   Q: Should pairing fail is the supplied points are not canonical?
+	   Q: Should pairing fail if there are no inputs?
+	*/
+
+	group := EcGroup(cx.program[cx.pc+1])
+	fs, ok := ecGroupSpecByField(group)
+	if !ok { // no version check yet, both appeared at once
+		return fmt.Errorf("invalid ec_pairing_check group %s", group)
+	}
+
 	last := len(cx.stack) - 1
 	prev := last - 1
-	aBytes := cx.stack[prev].Bytes
-	if len(aBytes) != 64 {
-		return errors.New("expect G1 in 64 bytes")
+	g1Bytes := cx.stack[prev].Bytes
+	g2Bytes := cx.stack[last].Bytes
+
+	var err error
+	ok = false
+	switch fs.field {
+	case BN254g2:
+		g1Bytes, g2Bytes = g2Bytes, g1Bytes
+		fallthrough
+	case BN254g1:
+		ok, err = bn254PairingCheck(g1Bytes, g2Bytes)
+	case BLS12_381g2:
+		g1Bytes, g2Bytes = g2Bytes, g1Bytes
+		fallthrough
+	case BLS12_381g1:
+		ok, err = bls12381PairingCheck(g1Bytes, g2Bytes)
+	default:
+		err = fmt.Errorf("invalid ec_pairing_check group %s", group)
 	}
-	a := bytesToBN254G1(aBytes)
-	kBytes := cx.stack[last].Bytes
-	k := new(big.Int).SetBytes(kBytes[:])
-	res := new(bn254.G1Affine).ScalarMultiplication(&a, k)
-	resBytes := bN254G1ToBytes(res)
+
 	cx.stack = cx.stack[:last]
-	cx.stack[prev].Bytes = resBytes
-	return nil
+	cx.stack[prev] = boolToSV(ok)
+	return err
```

Comment:
one question: if there is an erroring in pairing check, do we want to hide the erroring by casting existing err to an `ok = False`? I thought we might want to undistinguish following 2 cases:
- not pairing succeed
- pairing inputs are not valid

or in a sense, a failing pairing check should not fail the teal program?

Reply:
I think the intent is to fail the program since an error in the pairing check call is only from misformatted input, so it's probably programmer error. I guess most of the time, returning 0 would be ok (because most code will be written to "do nothing" when pairing fails), but it seems better to just reject the whole thing.

Reply:
yea that sounds fair to me, I just dug into the implementation, and realizing that the only erroring cases in our implementation and gnark implementation are related with wrong byte size (byte array len 0 or not aligned with 32 bytes), so I agree that it is falling in your category of wrong format input, which should fail anyways tho.